fish-rust/src/builtins/functions.rs

```diff
@@ -0,0 +1,413 @@
+            streams.err.append(wgettext_fmt!(
+                "%ls: Function '%ls' already exists. Cannot create copy '%ls'\n",
+                cmd,
+                new_func,
+                current_func
+            ));
+            builtin_print_error_trailer(parser, streams, cmd);
+            return STATUS_CMD_ERROR;
+        }
+        if function::copy(current_func, new_func.into(), parser) {
+            return STATUS_CMD_OK;
+        }
+        return STATUS_CMD_ERROR;
+    }
+
+    let mut res: c_int = STATUS_CMD_OK?;
+
+    let mut first = true;
+    for arg in args.iter() {
+        if opts.query {
+            if !function::exists(arg, parser) {
+                res += 1;
+            }
+        } else if let Some(props) = function::get_props_autoload(arg, parser) {
+            let path = props.definition_file().unwrap_or(L!(""));
+            let mut comment = WString::new();
+            if !first {
+                streams.out.append(L!("\n"));
+            };
+            if !opts.no_metadata {
+                if path.is_empty() {
+                    comment.push_utfstr(&wgettext_fmt!("Defined interactively"));
+                } else if path == "-" {
+                    comment.push_utfstr(&wgettext_fmt!("Defined via `source`"));
+                } else {
+                    let lineno: i32 = props.definition_lineno();
+                    comment.push_utfstr(&wgettext_fmt!("Defined in %ls @ line %d", path, lineno));
+                }
+
+                if props.is_copy() {
+                    let path = props.copy_definition_file().unwrap_or(L!(""));
+                    if path.is_empty() {
+                        comment.push_utfstr(&wgettext_fmt!(", copied interactively"));
+                    } else if path == "-" {
+                        comment.push_utfstr(&wgettext_fmt!(", copied via `source`"));
+                    } else {
+                        let lineno: i32 = props.copy_definition_lineno();
+                        comment.push_utfstr(&wgettext_fmt!(
+                            ", copied in %ls @ line %d",
+                            path,
+                            lineno
+                        ));
+                    }
+                }
+            }
+            let mut def = WString::new();
+
+            if !comment.is_empty() {
+                def.push_utfstr(&sprintf!(
+                    "# %ls\n%ls",
+                    comment,
+                    props.annotated_definition(arg)
+                ));
+            } else {
+                def = props.annotated_definition(arg);
+            }
+
+            if !streams.out_is_redirected && unsafe { isatty(STDOUT_FILENO) == 1 } {
+                let col = colorize_shell(&def.to_ffi(), parser.pin()).from_ffi();
+                streams.out.append(col);
+            } else {
+                streams.out.append(def);
+            }
+        } else {
+            res += 1;
+        }
+        first = false;
```

Comment:
(this always feels extremely clunky in any programming language for what is a very common thing - "only print a newline for every following entry")